pkg/ingester/index/index.go

```diff
@@ -0,0 +1,446 @@
+// originally from https://github.com/cortexproject/cortex/blob/868898a2921c662dcd4f90683e8b95c927a8edd8/pkg/ingester/index/index.go
+// but modified to support sharding queries.
+package index
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"sync"
+	"unsafe"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/querier/astmapper"
+)
+
+const indexShards = 32
+
+var ErrInvalidShardQuery = errors.New("incompatible index shard query")
+
+// InvertedIndex implements a in-memory inverter index from label pairs to fingerprints.
+// It is sharded to reduce lock contention on writes.
+type InvertedIndex struct {
+	totalShards uint32
+	shards      []*indexShard
+}
+
+// New returns a new InvertedIndex.
+func New() *InvertedIndex {
+	return NewWithShards(indexShards)
+}
+
+func NewWithShards(totalShards uint32) *InvertedIndex {
+	shards := make([]*indexShard, totalShards)
+	for i := uint32(0); i < totalShards; i++ {
+		shards[i] = &indexShard{
+			idx:   map[string]indexEntry{},
+			shard: i,
+		}
+	}
+	return &InvertedIndex{
+		totalShards: totalShards,
+		shards:      shards,
+	}
+}
+
+func (ii *InvertedIndex) getShards(shard *astmapper.ShardAnnotation) []*indexShard {
+	if shard == nil {
+		return ii.shards
+	}
+
+	indexFactor := int(ii.totalShards)
+	// calculate the start of the hash ring desired
+	lowerBound := shard.Shard * indexFactor / shard.Of
+	// calculate the end of the hash ring desired
+	upperBound := (shard.Shard + 1) * indexFactor / shard.Of
+	// see if the upper bound is cleanly doesn't align cleanly with the next shard
+	// which can happen when the schema sharding factor and inverted index
+	// sharding factor are not multiples of each other.
+	rem := (shard.Shard + 1) * indexFactor % shard.Of
+	if rem > 0 {
+		// there's overlap on the upper shard
+		upperBound = upperBound + 1
+	}
+
+	return ii.shards[lowerBound:upperBound]
+}
+
+func validateShard(totalShards uint32, shard *astmapper.ShardAnnotation) error {
```

Comment:
I'm not sure about what will happens if someone send 128 and use 8*16 the same shard. It could lock so I went for the safe way and only allows below the total shard queries. Currently it's always 16, so we can solve this later.